go/vt/vtgate/vindexes/vschema_routing_test.go

```diff
@@ -0,0 +1,500 @@
+		},
+	}
+	sharded2 := &vschemapb.Keyspace{
+		Sharded:                true,
+		RequireExplicitRouting: true,
+		Vindexes: map[string]*vschemapb.Vindex{
+			"xxhash": {
+				Type: "xxhash",
+			},
+		},
+		// none should be considered for choice as global or ambiguous because of RequireExplicitRouting
+		Tables: map[string]*vschemapb.Table{
+			"table6":   {}, // unique
+			"scommon2": {}, // common with sharded1, but has RequireExplicitRouting
+			"scommon3": {}, // unique
+		},
+	}
+	for _, ks := range []*vschemapb.Keyspace{sharded1, sharded2} {
+		for _, t := range ks.Tables {
+			t.ColumnVindexes = append(t.ColumnVindexes, &vschemapb.ColumnVindex{
+				Column: "c1",
+				Name:   "xxhash",
+			})
+		}
+	}
+	source := &vschemapb.SrvVSchema{
+		Keyspaces: map[string]*vschemapb.Keyspace{
+			"sharded1": sharded1,
+			"sharded2": sharded2,
+		},
+	}
+
+	vschema := BuildVSchema(source, sqlparser.NewTestParser())
+	require.NotNil(t, vschema)
+
+	// Check table is global
+	mustRouteGlobally := func(t *testing.T, tname, ksName string) {
+		t.Helper()
+		_, err := vschema.FindTable("", tname)
+		require.NoError(t, err)
+		// The vtgate data structures don't always set the keyspace name, so cannot reliably check that at the moment.
+		_ = ksName
+	}
+
+	mustNotRouteGlobally := func(t *testing.T, tname string) {
+		t.Helper()
+		_, err := vschema.FindTable("", tname)
+		require.Error(t, err)
+	}
+
+	// Verify the global tables
+	ks := vschema.Keyspaces["sharded1"]
+	require.EqualValues(t, vschema.globalTables, map[string]*Table{
+		"table5":   ks.Tables["table5"],
+		"scommon1": ks.Tables["scommon1"],
+		"scommon2": ks.Tables["scommon2"],
+	})
+	mustRouteGlobally(t, "table5", "sharded1")
+	mustRouteGlobally(t, "scommon1", "sharded1")
+	mustRouteGlobally(t, "scommon2", "sharded1")
+
+	// Add unsharded keyspaces to SrvVSchema and build VSchema
+	var err error
+	source.Keyspaces["unsharded1"] = unsharded1
+	source.Keyspaces["unsharded2"] = unsharded2
+	vschema.Keyspaces["unsharded1"], err = BuildKeyspace(unsharded1, sqlparser.NewTestParser())
+	require.NoError(t, err)
+	vschema.Keyspaces["unsharded2"], err = BuildKeyspace(unsharded2, sqlparser.NewTestParser())
+	require.NoError(t, err)
+
+	// Verify the global tables don't change
+	mustRouteGlobally(t, "table5", "sharded1")
+	mustRouteGlobally(t, "scommon1", "sharded1")
+	mustRouteGlobally(t, "scommon2", "sharded1")
+
+	// Add additional global tables and then verify that the unsharded global tables are added
+	AddAdditionalGlobalTables(source, vschema)
+
+	mustRouteGlobally(t, "table1", "unsharded1")
+	mustRouteGlobally(t, "table2", "unsharded1")
+
+	mustRouteGlobally(t, "table3", "unsharded2")
+	mustRouteGlobally(t, "table4", "unsharded2")
+	mustNotRouteGlobally(t, "ucommon3")
+
+	mustRouteGlobally(t, "scommon1", "sharded1")
+	mustRouteGlobally(t, "scommon2", "sharded1")
+	mustRouteGlobally(t, "table5", "sharded1")
+
+	mustNotRouteGlobally(t, "table6")
+	mustNotRouteGlobally(t, "scommon3")
+}
+
```

Comment:
The next test is moved over from `vschema_test.go` so that it is colocated with the new global routing tests.